File: package/android/src/main/java/com/margelo/nitro/nitromaps/MarkerSpatialIndex.kt
package com.margelo.nitro.nitromaps

import com.google.android.gms.maps.model.LatLngBounds

/**
 * Uniform grid spatial index over marker handles.
 *
 * Cells hold handles, not descriptors, and the grid is updated in place as
 * markers are inserted, moved and removed, so a moving marker costs one cell
 * swap instead of a rebuild. The grid bounds are computed over the dataset
 * with a margin; a marker that lands outside them flags a rebuild, which the
 * store runs once at the end of the batch that caused it.
 *
 * Not thread-safe on its own: the owning [MarkerStore] serializes access.
 */
internal class MarkerSpatialIndex(cellsPerSide: Int = 96) {
  private val side: Int = maxOf(1, cellsPerSide)
  private var minLat = 0.0
  private var maxLat = 0.0
  private var minLon = 0.0
  private var maxLon = 0.0
  private var latStep = 1.0
  private var lonStep = 1.0
  private var hasBounds = false
  private var needsRebuild = false
  private val cells: Array<IntList> = Array(side * side) { IntList() }
  /** Cell index per handle, -1 when the handle is not indexed. */
  private var cellOf = IntArray(0)
  var count = 0
    private set

  fun insert(handle: Int, latitude: Double, longitude: Double) {
    ensureCapacity(handle)
    if (cellOf[handle] >= 0) {
      move(handle, latitude, longitude)
      return
    }

    count += 1
    if (!hasBounds || !contains(latitude, longitude)) {
      needsRebuild = true
    }
    val cell = clampedCellIndex(latitude, longitude)
    cells[cell].add(handle)
    cellOf[handle] = cell
  }

  fun move(handle: Int, latitude: Double, longitude: Double) {
    if (handle >= cellOf.size || cellOf[handle] < 0) {
      insert(handle, latitude, longitude)
      return
    }

    if (!contains(latitude, longitude)) {
      needsRebuild = true
    }
    val current = cellOf[handle]
    val next = clampedCellIndex(latitude, longitude)
    if (next == current) {
      return
    }
    cells[current].removeValue(handle)
    cells[next].add(handle)
    cellOf[handle] = next
  }

  fun remove(handle: Int) {
    if (handle >= cellOf.size || cellOf[handle] < 0) {
      return
    }
    cells[cellOf[handle]].removeValue(handle)
    cellOf[handle] = -1
    count -= 1
  }

  fun removeAll() {
    cells.forEach { it.clear() }
    cellOf = IntArray(0)
    count = 0
    hasBounds = false
    needsRebuild = false
  }

  /**
   * Recomputes the grid over every live marker if one fell outside the current
   * bounds. Called once per applied batch, before any query.
   */
  fun rebuildIfNeeded(latitudes: DoubleArray, longitudes: DoubleArray, flags: ByteArray) {
    if (!needsRebuild) {
      return
    }
    needsRebuild = false

    var minLatV = Double.MAX_VALUE
    var maxLatV = -Double.MAX_VALUE
    var minLonV = Double.MAX_VALUE
    var maxLonV = -Double.MAX_VALUE
    var alive = 0
    for (handle in flags.indices) {
      if (flags[handle].toInt() and MarkerStore.FLAG_ALIVE == 0) continue
      alive += 1
      if (latitudes[handle] < minLatV) minLatV = latitudes[handle]
      if (latitudes[handle] > maxLatV) maxLatV = latitudes[handle]
      if (longitudes[handle] < minLonV) minLonV = longitudes[handle]
      if (longitudes[handle] > maxLonV) maxLonV = longitudes[handle]
    }

    cells.forEach { it.clear() }
    if (alive == 0) {
      hasBounds = false
      cellOf.fill(-1)
      count = 0
      return
    }

    // A margin keeps ordinary movement inside the grid; only a marker that
    // leaves the dataset's neighbourhood triggers the next rebuild.
    val latPad = maxOf((maxLatV - minLatV) * 0.15, 1e-6)
    val lonPad = maxOf((maxLonV - minLonV) * 0.15, 1e-6)
    minLat = minLatV - latPad
    maxLat = maxLatV + latPad
    minLon = minLonV - lonPad
    maxLon = maxLonV + lonPad
    latStep = maxOf(1e-9, (maxLat - minLat) / side)
    lonStep = maxOf(1e-9, (maxLon - minLon) / side)
    hasBounds = true

    ensureCapacity(flags.size - 1)
    for (handle in flags.indices) {
      if (flags[handle].toInt() and MarkerStore.FLAG_ALIVE != 0) {
        val cell = clampedCellIndex(latitudes[handle], longitudes[handle])
        cells[cell].add(handle)
        cellOf[handle] = cell
      } else {
        cellOf[handle] = -1
      }
    }
    count = alive
  }

  /** Handles whose grid cells overlap the padded bounds. */
  fun candidates(bounds: LatLngBounds, padding: Double = 0.2): IntArray {
    if (count == 0 || !hasBounds) {
      return IntArray(0)
    }

    val latSpan = bounds.northeast.latitude - bounds.southwest.latitude
    val lonSpan = if (bounds.northeast.longitude < bounds.southwest.longitude) {
      bounds.northeast.longitude - bounds.southwest.longitude + 360.0
    } else {
      bounds.northeast.longitude - bounds.southwest.longitude
    }
    val latPad = latSpan * padding
    val lonPad = lonSpan * padding
    val minLatQ = bounds.southwest.latitude - latPad
    val maxLatQ = bounds.northeast.latitude + latPad
    if (maxLatQ < minLat || minLatQ > maxLat) {
      return IntArray(0)
    }

    val minLonQ = bounds.southwest.longitude - lonPad
    val maxLonQ = bounds.northeast.longitude + lonPad
    if (!overlapsLongitude(minLonQ, maxLonQ)) {
      return IntArray(0)
    }

    val rowStart = clampedRow(minLatQ)
    val rowEnd = clampedRow(maxLatQ)
    val result = IntList(64)
    val columns = longitudeColumns(minLonQ, maxLonQ)
    var row = rowStart
    while (row <= rowEnd) {
      val base = row * side
      for (column in columns) {
        result.addAll(cells[base + column])
      }
      row += 1
    }
    return result.toIntArray()
  }

  private fun ensureCapacity(handle: Int) {
    if (handle < cellOf.size) {
      return
    }
    val previous = cellOf.size
    cellOf = cellOf.copyOf(maxOf(handle + 1, previous * 2, 64))
    cellOf.fill(-1, previous, cellOf.size)
  }

  /**
   * Whether a query's longitude range, which may cross the antimeridian, meets
   * the grid's. Without this a query east or west of the dataset would clamp to
   * the outermost column and return everything in it.
   */
  private fun overlapsLongitude(minLonQ: Double, maxLonQ: Double): Boolean {
    if (maxLonQ - minLonQ >= 360.0) {
      return true
    }
    val wrappedMin = wrapLongitude(minLonQ)
    val wrappedMax = wrapLongitude(maxLonQ)
    return if (wrappedMin <= wrappedMax) {
      wrappedMax >= minLon && wrappedMin <= maxLon
    } else {
      maxLon >= wrappedMin || minLon <= wrappedMax
    }
  }

  private fun contains(latitude: Double, longitude: Double): Boolean {
    return latitude >= minLat && latitude <= maxLat && longitude >= minLon && longitude <= maxLon
  }

  private fun longitudeColumns(minLon: Double, maxLon: Double): List<Int> {
    if (maxLon - minLon >= 360.0) {
      return (0 until side).toList()
    }

    val wrappedMin = wrapLongitude(minLon)
    val wrappedMax = wrapLongitude(maxLon)
    if (wrappedMin <= wrappedMax && maxLon <= 180.0 && minLon >= -180.0) {
      return (clampedColumn(wrappedMin)..clampedColumn(wrappedMax)).toList()
    }

    val firstRange = clampedColumn(wrappedMin) until side
    val secondRange = 0..clampedColumn(wrappedMax)
    return firstRange.toList() + secondRange.toList()
  }

  private fun wrapLongitude(lon: Double): Double {
    var wrapped = lon
    while (wrapped > 180.0) wrapped -= 360.0
    while (wrapped < -180.0) wrapped += 360.0
    return wrapped
  }

  private fun clampedCellIndex(lat: Double, lon: Double): Int {
    return clampedRow(lat) * side + clampedColumn(lon)
  }

  private fun clampedRow(lat: Double): Int {
    return minOf(side - 1, maxOf(0, ((lat - minLat) / latStep).toInt()))
  }

  private fun clampedColumn(lon: Double): Int {
    return minOf(side - 1, maxOf(0, ((lon - minLon) / lonStep).toInt()))
  }
}

File: package/ios/MarkerSpatialIndex.swift
import MapKit

/// Uniform grid spatial index over marker handles.
///
/// Cells hold handles, not descriptors, and the grid is updated in place as
/// markers are inserted, moved and removed, so a moving marker costs one cell
/// swap instead of a rebuild. The grid bounds are computed over the dataset
/// with a margin; a marker that lands outside them flags a rebuild, which the
/// store runs once at the end of the batch that caused it.
///
/// Not thread-safe on its own: the owning `MarkerStore` serializes access.
final class MarkerSpatialIndex {
  private let cellsPerSide: Int
  private var minLat = 0.0
  private var maxLat = 0.0
  private var minLon = 0.0
  private var maxLon = 0.0
  private var latStep = 1.0
  private var lonStep = 1.0
  private var hasBounds = false
  private var needsRebuild = false
  private var cells: [[Int32]]
  /// Cell index per handle, -1 when the handle is not indexed.
  private var cellOf: [Int32] = []
  private(set) var count = 0

  init(cellsPerSide: Int = 96) {
    let side = max(1, cellsPerSide)
    self.cellsPerSide = side
    cells = Array(repeating: [], count: side * side)
  }

  func insert(_ handle: Int, latitude: Double, longitude: Double) {
    ensureCapacity(handle)
    guard cellOf[handle] < 0 else {
      move(handle, latitude: latitude, longitude: longitude)
      return
    }

    count += 1
    if !hasBounds || !contains(latitude: latitude, longitude: longitude) {
      needsRebuild = true
    }
    let cell = clampedCellIndex(latitude: latitude, longitude: longitude)
    cells[cell].append(Int32(handle))
    cellOf[handle] = Int32(cell)
  }

  func move(_ handle: Int, latitude: Double, longitude: Double) {
    guard handle < cellOf.count, cellOf[handle] >= 0 else {
      insert(handle, latitude: latitude, longitude: longitude)
      return
    }

    if !contains(latitude: latitude, longitude: longitude) {
      needsRebuild = true
    }
    let current = Int(cellOf[handle])
    let next = clampedCellIndex(latitude: latitude, longitude: longitude)
    guard next != current else {
      return
    }
    removeFromCell(handle, cell: current)
    cells[next].append(Int32(handle))
    cellOf[handle] = Int32(next)
  }

  func remove(_ handle: Int) {
    guard handle < cellOf.count, cellOf[handle] >= 0 else {
      return
    }
    removeFromCell(handle, cell: Int(cellOf[handle]))
    cellOf[handle] = -1
    count -= 1
  }

  func removeAll() {
    for index in cells.indices {
      cells[index].removeAll(keepingCapacity: false)
    }
    cellOf.removeAll()
    count = 0
    hasBounds = false
    needsRebuild = false
  }

  /// Recomputes the grid over every live marker if one fell outside the
  /// current bounds. Called once per applied batch, before any query.
  func rebuildIfNeeded(latitudes: [Double], longitudes: [Double], flags: [UInt8]) {
    guard needsRebuild else {
      return
    }
    needsRebuild = false

    var minLatV = Double.greatestFiniteMagnitude
    var maxLatV = -Double.greatestFiniteMagnitude
    var minLonV = Double.greatestFiniteMagnitude
    var maxLonV = -Double.greatestFiniteMagnitude
    var alive = 0
    for handle in 0..<flags.count where flags[handle] & MarkerStore.Flag.alive != 0 {
      alive += 1
      minLatV = min(minLatV, latitudes[handle])
      maxLatV = max(maxLatV, latitudes[handle])
      minLonV = min(minLonV, longitudes[handle])
      maxLonV = max(maxLonV, longitudes[handle])
    }

    for index in cells.indices {
      cells[index].removeAll(keepingCapacity: true)
    }
    guard alive > 0 else {
      hasBounds = false
      for handle in cellOf.indices {
        cellOf[handle] = -1
      }
      count = 0
      return
    }

    // A margin keeps ordinary movement inside the grid; only a marker that
    // leaves the dataset's neighbourhood triggers the next rebuild.
    let latPad = max((maxLatV - minLatV) * 0.15, 1e-6)
    let lonPad = max((maxLonV - minLonV) * 0.15, 1e-6)
    minLat = minLatV - latPad
    maxLat = maxLatV + latPad
    minLon = minLonV - lonPad
    maxLon = maxLonV + lonPad
    latStep = max(1e-9, (maxLat - minLat) / Double(cellsPerSide))
    lonStep = max(1e-9, (maxLon - minLon) / Double(cellsPerSide))
    hasBounds = true

    ensureCapacity(flags.count - 1)
    for handle in 0..<flags.count {
      if flags[handle] & MarkerStore.Flag.alive != 0 {
        let cell = clampedCellIndex(latitude: latitudes[handle], longitude: longitudes[handle])
        cells[cell].append(Int32(handle))
        cellOf[handle] = Int32(cell)
      } else if handle < cellOf.count {
        cellOf[handle] = -1
      }
    }
    count = alive
  }

  /// Handles whose grid cells overlap the padded region.
  func candidates(in region: MKCoordinateRegion, padding: Double = 0.2) -> [Int32] {
    guard count > 0, hasBounds else {
      return []
    }

    let latPad = region.span.latitudeDelta * padding
    let lonPad = region.span.longitudeDelta * padding
    let minLatQ = region.center.latitude - region.span.latitudeDelta / 2 - latPad
    let maxLatQ = region.center.latitude + region.span.latitudeDelta / 2 + latPad
    let minLonQ = region.center.longitude - region.span.longitudeDelta / 2 - lonPad
    let maxLonQ = region.center.longitude + region.span.longitudeDelta / 2 + lonPad
    guard maxLatQ >= minLat, minLatQ <= maxLat else {
      return []
    }

    let rowStart = clampedRow(minLatQ)
    let rowEnd = clampedRow(maxLatQ)
    let columns = longitudeColumnRange(minLon: minLonQ, maxLon: maxLonQ)

    var result: [Int32] = []
    var row = rowStart
    while row <= rowEnd {
      let base = row * cellsPerSide
      for column in columns {
        result.append(contentsOf: cells[base + column])
      }
      row += 1
    }
    return result
  }

  private func ensureCapacity(_ handle: Int) {
    guard handle >= cellOf.count else {
      return
    }
    cellOf.append(contentsOf: repeatElement(-1, count: handle + 1 - cellOf.count))
  }

  private func removeFromCell(_ handle: Int, cell: Int) {
    guard let position = cells[cell].firstIndex(of: Int32(handle)) else {
      return
    }
    cells[cell].swapAt(position, cells[cell].count - 1)
    cells[cell].removeLast()
  }

  private func contains(latitude: Double, longitude: Double) -> Bool {
    latitude >= minLat && latitude <= maxLat && longitude >= minLon && longitude <= maxLon
  }

  private func longitudeColumnRange(minLon: Double, maxLon: Double) -> [Int] {
    if maxLon - minLon >= 360.0 {
      return Array(0..<cellsPerSide)
    }

    let wrappedMin = wrapLongitude(minLon)
    let wrappedMax = wrapLongitude(maxLon)
    if wrappedMin <= wrappedMax && maxLon <= 180.0 && minLon >= -180.0 {
      let colStart = clampedColumn(wrappedMin)
      let colEnd = clampedColumn(wrappedMax)
      return Array(colStart...colEnd)
    }

    let firstRange = clampedColumn(wrappedMin)...cellsPerSide - 1
    let secondRange = 0...clampedColumn(wrappedMax)
    return Array(firstRange) + Array(secondRange)
  }

  private func wrapLongitude(_ lon: Double) -> Double {
    var wrapped = lon
    while wrapped > 180.0 { wrapped -= 360.0 }
    while wrapped < -180.0 { wrapped += 360.0 }
    return wrapped
  }

  private func clampedCellIndex(latitude: Double, longitude: Double) -> Int {
    clampedRow(latitude) * cellsPerSide + clampedColumn(longitude)
  }

  private func clampedRow(_ lat: Double) -> Int {
    min(cellsPerSide - 1, max(0, Int((lat - minLat) / latStep)))
  }

  private func clampedColumn(_ lon: Double) -> Int {
    min(cellsPerSide - 1, max(0, Int((lon - minLon) / lonStep)))
  }
}
